### PR TITLE
Big refactor for updated JsonProtocol

### PR DESCRIPTION
**Pull request by ddd-mtl**
(works with https://github.com/holochain/holochain-rust/pull/1457)
Removed hackmode and IPC-MOCK mode.
realmem doesnt really store anything. Network module will ask Core for data.
Moved somme bookkeeping to N3hMode.
Now have one dht per Chain, and not per dna.
entryAspect concept...

### Open questions

- Should dht backend broadcast entry to all known agents on publishEntry?
- Do we keep having Mem store data?
- Should MemoryServer gossip like n3h?

### CODE REVIEW

File: lib/hackmode/realmem.test.js [hunk 1]
```diff
@@ -1,30 +1,112 @@
...
+    }
+    ])
+  })
+
+  it('should have all entryAddress', () => {
+    expect(m.has('entryAddress1', 'aspectAddress1')).equals(true)
+    expect(m.has('entryAddress2', 'aspectAddress1')).equals(false)
+    expect(m.has('entryAddress2', 'aspectAddress2')).equals(true)
+    expect(m.has('entryAddress2', 'aspectAddress3')).equals(true)
+    expect(m.has('entryAddress2', 'aspectAddress1')).equals(false)
```
Comment (jmday) [inline lib/hackmode/realmem.test.js]: This line is redundant with line 60 above. Consider:
`    expect(m.has('entryAddress1', 'aspectAddress2')).equals(false)`
`    expect(m.has('entryAddress1', 'aspectAddress3')).equals(false)`

File: lib/hackmode/realmem.test.js [hunk 2]
```diff
@@ -1,30 +1,112 @@
...
+      address: 'aspectAddress2'
+    })
+    aspect = m.getAspect('entryAddress2', 'aspectAddress3')
+    expect(aspect).deep.equals({
+      entryAddress: 'entryAddress2',
+      address: 'aspectAddress3'
+    })
+  })
+
+  it('should not insert', () => {
```
Comment (jmday) [inline lib/hackmode/realmem.test.js]: consider:
`  it('should not insert duplicate', () => {`

File: lib/hackmode/realmode.js [hunk 3]
```diff
@@ -284,143 +274,84 @@ class N3hRealMode extends N3hMode {
...
+      case 'handleQueryEntryResult':
+        // Note: data is a QueryEntryResultData
+        if (!this._hasTrackOrFail(data.responderAgentId, data.dnaAddress, data._id)) {
+          return
+        }
+        // #fullsync hack: send response back to self (ipc)
+        data.method = 'queryEntryResult'
         this._ipcSend('json', data)
         return
       case 'handleFetchEntryResult':
```
Comment (jmday) [inline lib/hackmode/realmode.js]: Should this be handleQueryEntryResult now?

File: lib/hackmode/realmode.js [hunk 4]
```diff
@@ -171,8 +171,8 @@ class N3hRealMode extends N3hMode {
     log.i(this.me() + 'p2p bound', this._p2p.getAdvertise())
   }
 
-  async _sendToAgent (msg, dnaAddress, toAgentId, fromAgentId, reqId) {
-    const mId = this._getPeerAddressOrFail(dnaAddress, toAgentId, fromAgentId, reqId)
+  async _sendToAgent (msg, dnaAddress, agentId, toAgentId, fromAgentId, reqId) {
```
Comment (jmday) [inline lib/hackmode/realmode.js]: If the `agentId` here is referring to the agentId associated with the chainId, might this be called `chainAgentId` for clarity?

File: lib/hackmode/realmode.js [hunk 5]
```diff
@@ -791,65 +821,70 @@ class N3hRealMode extends N3hMode {
...
    *  Check if agent is tracking dna.
    *  If not, will try to send a FailureResult back to sender (if sender info is provided).
    *  Returns peerAddress of receiverAgentId if agent is tracking dna.
    *  @private
+   *  FIXME
    */
-  _getPeerAddressOrFail (dnaAddress, receiverAgentId, senderAgentId, requestId) {
-    const recvPeer = this._getDhtRef(dnaAddress).getPeerLocal(receiverAgentId)
-    if (!recvPeer) {
+  _getPeerAddressOrFail (dnaAddress, agentId, receiverAgentId, senderAgentId, requestId) {
```
Comment (jmday) [inline lib/hackmode/realmode.js]: suggest changing `agentId` to `chainAgentId`

File: lib/hackmode/realmode.js (hunk 4 above)
Comment (jmday) [inline lib/hackmode/realmode.js]: suggest changing `agentId` to `chainAgentId`

File: lib/hackmode/realmode.js (hunk 5 above)
Comment (jmday) [reply lib/hackmode/realmode.js]: or perhaps `providerAgentId`?

Comment (ddd-mtl) [reply lib/hackmode/realmode.js]: outdated code

File: lib/hackmode/realmode.js (hunk 4 above)
Comment (ddd-mtl) [reply lib/hackmode/realmode.js]: outdated code

File: lib/hackmode/realmem.test.js (hunk 1 above)
Comment (ddd-mtl) [reply lib/hackmode/realmem.test.js]: 👍

File: lib/hackmode/realmem.test.js (hunk 2 above)
Comment (ddd-mtl) [reply lib/hackmode/realmem.test.js]: 👍

File: lib/hackmode/realmode.js (hunk 3 above)
Comment (ddd-mtl) [reply lib/hackmode/realmode.js]: ???
`handleQueryEntryResult` is handled just above.
`handleFetchEntryResult` is for resolving a data request initiated by gossip

**Comment by neonphog** [comment]
> ### Open questions
> - Should dht backend broadcast entry to all known agents on publishEntry?

Yes.

> - Do we keep having Mem store data?

I think we should go ahead and stop storing data. This may mean we need to do some extra testing / debugging with core to make sure the core side is working with this before we merge

> - Should MemoryServer gossip like n3h?

We've got a lot of thinking still to be done about gossip... maybe path of least resistance for the time-being?

**Pull request merged**